### DISCUSSION TITLE
use custom input iterator for serialization

### DISCUSSION THREAD

**Discussion by dhebbeker** [nlohmann/json#4355] [Q&A]
I have data in an incompatible, proprietary container which I would like to serialize. I thought of using a custom iterator which would use that container and provide data in a compatible/serializable format. That is, de-referencing the iterator would convert the respective element on-demand.

I hoped I could simply pass `begin` and `end` iterator to the `json`  constructor. But when I tried this with a well defined container, compilation fails:

```c++
#include <nlohmann/json.hpp>
#include <vector>
#include <iterator>

int main() {
  std::vector<int> v;
  nlohmann::json j(std::begin(v), std::end(v));
}
```

Is there a way to achieve this?

**Comment by nlohmann** [comment]
Do you have more details like the source of your container or the error message?

**Reply by dhebbeker** [reply]
I have a `T1` of `std::map` which can not be automatically serialized by `json`. But I have an input iterator type which is constructed with a `T1::iterator` and which has a `T2 operator*()` overload. `T2` can be serialized through a custom `to_json()`.

The code I am working on is public, if you really want to dig into the [details](https://github.com/Task-Tracker-Systems/Task-Tracker-Device/blob/af269f3549761402113dcf2a83dc1649948b6141/lib/3rd_party_adapters/nlohmann/JsonGenerator.cpp#L42) (WIP!). 

As I was not succeeding, I though of trying with a simplified example (see [above](https://github.com/nlohmann/json/discussions/4355#discussion-6541502)). Checking if it works with C++ Standard Library iterators. But this fails with clang 15 (see [full output on Compiler Explorer](https://godbolt.org/z/4YnzexMGh)). I think the relevant error message is:

```
<source>:7:18: error: no matching constructor for initialization of 'nlohmann::json' (aka 'basic_json<>')
  nlohmann::json j(std::begin(v), std::end(v));
                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~
...
/opt/compiler-explorer/libs/nlohmann_json/trunk/single_include/nlohmann/json.hpp:20335:5: note: candidate template ignored: requirement 'std::is_same<__gnu_cxx::__normal_iterator<int *, std::vector<int, std::allocator<int>>>, nlohmann::detail::iter_impl<nlohmann::basic_json<std::map, std::vector, std::basic_string<char, std::char_traits<char>, std::allocator<char>>, bool, long, unsigned long, double, std::allocator, nlohmann::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char>>, void>>>::value || std::is_same<__gnu_cxx::__normal_iterator<int *, std::vector<int, std::allocator<int>>>, nlohmann::detail::iter_impl<const nlohmann::basic_json<std::map, std::vector, std::basic_string<char, std::char_traits<char>, std::allocator<char>>, bool, long, unsigned long, double, std::allocator, nlohmann::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char>>, void>>>::value' was not satisfied [with InputIT = __gnu_cxx::__normal_iterator<int *, std::vector<int>>]
    basic_json(InputIT first, InputIT last)
```

**Comment by dhebbeker** [comment]
I was hoping to use constructor overload number 7 from the documentation: https://json.nlohmann.me/api/basic_json/basic_json/
